packages/next/src/server/app-render/fast-set-immediate.external.ts

```diff
@@ -0,0 +1,531 @@
+import { InvariantError } from '../../shared/lib/invariant-error'
+import { bindSnapshot } from './async-local-storage'
+
+enum ExecutionState {
+  None = 1,
+  Waiting = 2,
+  Working = 3,
+}
+
+let isInstalled = false
+let wasEnabledAtLeastOnce = false
+
+const queuedImmediates: QueueItem[] = []
+let pendingNextTicks = 0
+let executionState: ExecutionState = ExecutionState.None
+
+const originalSetImmediate = globalThis.setImmediate
+const originalClearImmediate = globalThis.clearImmediate
+const originalNextTick = process.nextTick
+
+export function install() {
+  if (process.env.NEXT_RUNTIME === 'edge') {
+    throw new InvariantError(
+      'Fast setImmediate is not available in the edge runtime.'
+    )
+  } else {
+    const nodeTimers = require('node:timers') as typeof import('node:timers')
+    globalThis.setImmediate = nodeTimers.setImmediate =
+      // Workaround for missing __promisify__ which is not a real property
+      patchedSetImmediate as unknown as typeof setImmediate
+    globalThis.clearImmediate = nodeTimers.clearImmediate =
+      patchedClearImmediate
+
+    const nodeTimersPromises =
+      require('node:timers/promises') as typeof import('node:timers/promises')
+    nodeTimersPromises.setImmediate =
+      patchedSetImmediatePromise as typeof import('node:timers/promises').setImmediate
+
+    process.nextTick = patchedNextTick
+
+    isInstalled = true
+  }
+}
+
+/**
+ * **WARNING: This function changes the usual behavior of the event loop!**
+ * **Be VERY careful about where you call it.**
+ *
+ * Starts capturing calls to `setImmediate` to run them as "fast immediates".
+ * All calls captured in this way will be executed after the current task
+ * (after ticks from `process.nextTick()` and microtasks scheduled from ticks).
+ * This function needs to be called again in each task that needs the
+ * "fast immediates" behavior.
+ *
+ * ### Motivation
+ *
+ * We don't want `setImmediate` to be considered IO in Cache Components.
+ * To achieve this in a staged (pre)render, we want to allow immediates scheduled
+ * in stage N to run before stage N+1.
+ * Since we schedule stages using sequential `setTimeout`, this isn't possible without
+ * intercepting `setImmediate` and doing the scheduling on our own.
+ * We refer to this as a "fast immediate".
+ *
+ * Notably, this affects React's `scheduleWork` in render, which uses `setImmediate`.
+ * This is desirable -- if async work was scheduled during a stage, then it should
+ * get to run before we finish that stage.
+ *
+ * ### Example
+ *
+ * ```ts
+ * setTimeout(() => {
+ *   runPendingImmediatesAfterCurrentTask()
+ *   console.log("timeout 1")
+ *   setImmediate(() => {
+ *     console.log("immediate!!!")
+ *   })
+ * })
+ * setTimeout(() => {
+ *   console.log("timeout 1")
+ * })
+ * ```
+ * will print
+ *
+ * ```
+ * timeout 1
+ * immediate!!!
+ * timeout 2
+ * ```
+ *
+ * instead of the normal order
+ * ```
+ * timeout 1
+ * timeout 2
+ * immediate!!!
+ * ```
```

Comment:
We should clarify this as usual order and remind our future selves that Node does not guarantee this order and it can run the immediate before or after timeout2 depending on a variety of runtime conditions that are indiscernable to the user